pkg/metadata/host/host_tags.go

```diff
@@ -106,39 +113,37 @@ func GetHostTags(ctx context.Context, cached bool) *Tags {
 		return nil, nil
 	}
 
-	providers := make(map[string]*providerDef)
-
-	if config.Datadog.GetBool("collect_ec2_tags") {
-		// WARNING: if this config is enabled on a non-ec2 host, then its
-		// retries may time out, causing a 3s delay
-		providers["ec2"] = &providerDef{10, ec2.GetTags, false}
-	}
-
+	providers := getProvidersDefinitionsFunc()
+	// GCE is handled separately as it requires inlined var/function
 	if config.Datadog.GetBool("collect_gce_tags") {
 		providers["gce"] = &providerDef{1, getGCE, false}
 	}
 
-	if config.IsFeaturePresent(config.Kubernetes) {
-		providers["kubernetes"] = &providerDef{10, k8s.GetTags, false}
-	}
-
-	if config.IsFeaturePresent(config.Docker) {
-		providers["docker"] = &providerDef{1, docker.GetTags, false}
-	}
-
 	for {
 		for name, provider := range providers {
 			provider.retries--
+			providerCacheKey := buildKey("provider-" + name)
 			tags, err := provider.getTags(ctx)
-			if err != nil {
-				log.Debugf("No %s host tags, remaining attempts: %d, err: %v", name, provider.retries, err)
-			} else {
-				provider.retrieved = true
+			if err == nil {
+				if provider.cacheFallback {
+					cache.Cache.Set(providerCacheKey, tags, cache.NoExpiration)
```

Comment:
I'm curious why use a process-global cache, rather than just a local map?

Reply:
As the function is not attached to any object, it would need to be a package variable that we need to lock.
I believe using `cache.Cache` is more common in the Agent, so using it to keep the code consistent.